auto_submit/lib/requests/check_pull_request.dart

```diff
@@ -88,6 +105,151 @@ class CheckPullRequest extends RequestHandler {
         isConflicting: isConflicting,
         unknownMergeableState: unknownMergeableState);
   }
+
+  /// Returns whether all statuses are successful.
+  ///
+  /// Also fills [failures] with the names of any status/check that has failed.
+  Future<bool> _checkStatuses(
+    RepositorySlug slug,
+    String sha,
+    Set<_FailureDetail> failures,
+    List<RepositoryStatus> statuses,
+    List<CheckRun> checkRuns,
+    CheckSuite? checkSuite,
+    String name,
+    List<String> labels,
+  ) async {
+    assert(failures.isEmpty);
+    bool allSuccess = true;
+
+    // The status checks that are not related to changes in this PR.
+    const Set<String> notInAuthorsControl = <String>{
+      'luci-flutter', // flutter repo
+      'luci-engine', // engine repo
+      'submit-queue', // plugins repo
+    };
+
+    // Ensure repos with tree statuses have it set
+    if (Config.reposWithTreeStatus.contains(slug)) {
+      bool treeStatusExists = false;
+      final String treeStatusName = 'luci-${slug.name}';
+
+      // Scan list of statuses to see if the tree status exists (this list is expected to be <5 items)
+      for (RepositoryStatus status in statuses) {
+        if (status.context == treeStatusName) {
+          treeStatusExists = true;
+        }
+      }
+
+      if (!treeStatusExists) {
+        failures.add(_FailureDetail('tree status $treeStatusName', 'https://flutter-dashboard.appspot.com/#/build'));
+      }
+    }
+
+    final String overrideTreeStatusLabel = config.overrideTreeStatusLabel;
+    log.info('Validating name: $name, status: $statuses');
+    for (RepositoryStatus status in statuses) {
+      final String? name = status.context;
+      if (status.state != 'success') {
+        if (notInAuthorsControl.contains(name) && labels.contains(overrideTreeStatusLabel)) {
+          continue;
+        }
+        allSuccess = false;
+        if (status.state == 'failure' && !notInAuthorsControl.contains(name)) {
+          failures.add(_FailureDetail(name!, status.targetUrl as String));
+        }
+      }
+    }
+
+    log.info('Validating name: $name, checks: $checkRuns');
+    for (CheckRun checkRun in checkRuns) {
+      final String? name = checkRun.name;
+      if (checkSuite!.conclusion == CheckRunConclusion.success) {
+        continue;
+      } else if (checkRun.status == CheckRunStatus.completed) {
+        failures.add(_FailureDetail(name!, checkRun.detailsUrl as String));
+      }
+      allSuccess = false;
+    }
+
+    // Validate cirrus
+    const List<String> _failedStates = <String>['FAILED', 'ABORTED'];
+    const List<String> _succeededStates = <String>['COMPLETED', 'SKIPPED'];
+
+    //TODO(Kristin): Distinguish check runs from cirrus or flutter-dashboard. https://github.com/flutter/flutter/issues/99805.
```

Comment:
Let's put this comment above line 165: `for (CheckRun checkRun in checkRuns) {`, and remove all below cirrus logics.

The distinction part should happen inside the checkrun loop.

Reply:
Sure, done.

Reply:
Yes. These logics will be obsolete if we switch the check results from cirrus GraphQL to Rest API.

I would recommend to add cirrus related logics when solving: https://github.com/flutter/flutter/issues/99805, where you only need to distinguish the FailureDetail. The check run result check has been taken care of in the line 165 loop.

Reply:
Gotcha Keyong, thanks. I will remove them here and add related logics later in the correct places.